### DISCUSSION TITLE
Drop support for encoded data

### DISCUSSION THREAD

**Discussion by PGijsbers** [openml/automlbenchmark#663] [Ideas]
I am considering to drop the support for encoded data. When we started this project in 2018, there were still several AutoML frameworks which could not handle categorical and/or text data as part of tabular data natively. To that end, we also supported providing the frameworks with data-encoded all-numerical numpy arrays. However, nowadays almost all frameworks support the inclusion of text data.

Going into 2025, I think it's completely fair to expect AutoML frameworks to 1) be able to take columns with text data as input and 2) be able to take textual class labels as input and produce them as predictions. For that reason, and to reduce the complexity of some of the data loading and serialization code (and consequently the options exposed to integration scripts), I propose to do away with the option to get encoded data directly from the benchmark software.

The drawback is that we may need to drop support for some frameworks which have ceased to be updated (Auto-WEKA, TPOT), and/or provide a legacy pipeline (also for e.g., the baselines). On the balance, I think this is very reasonable. In the case these old systems would need to be benchmarked, older versions of the AutoML benchmark may continue to be used.

I hope people in the AutoML community can weigh in on this decision.

(in particular because of earlier contributions to the benchmark, or because they have a framework which currently uses encoded data, I am also pinging @mfeurer @eddiebergman @LennartPurucker @Innixma @shchur @fmohr - but the discussion is open to everyone )

**Comment by eddiebergman** [comment]
Hiyo, I think that's perfectly reasonable and I often use AutoMLBenchmark as a test for handling all the subtleties of non-encoded data. With regards to backwards compatibility, I would suggest that perhaps `amlb` exposes some default encoder (what's there currently?) that integrating frameworks can utilize in their own `def run()` if they like. One issue with this is that the environment in which the framework is run will not have `amlb` installed, perhaps either it's a relative import or at the very least, a copy-pastable function. Either way, responsibility for this is best left to the implementor of a framework.

**Reply by PGijsbers** [reply]
That's kind of the way it currently works. There is a [custom encoder](https://github.com/openml/automlbenchmark/blob/aded7c7ada67fb2b7871cbd2b461a0fb93ec8baa/amlb/datautils.py#L162) in the datautils directory, which is just a wrapper around a set of scikit-learn encoders on how to deal with missing values. Though this is often also used under the hood by e.g., `data.X_enc`. However, the modern scikit-learn encoders might be able to replace this encoder entirely, I would have to investigate. If I would leave this in, I would probably prefer for people to call the encoder explicitly only.

**Reply by mfeurer** [reply]
How about using skrub as a default encoder?

**Comment by LennartPurucker** [comment]
+1. In my opinion, even the base models (think XGBoost, CatBoost, TabPFN, ...) of AutoML should be able to handle the raw data starting in 2025. 

Even if scikit-learn does not follow this approach.

**Reply by PGijsbers** [reply]
I would agree, since the type of encoding matters w.r.t. the different base models.

**Reply by mfeurer** [reply]
Can they handle raw strings?

**Reply by LennartPurucker** [reply]
No, but the framework / individual model should then implement code that handles any data as intended by the developer of the framework / individual model.

**Reply by LennartPurucker** [reply]
Oh, CatBoost can AFAIK.

**Reply by Innixma** [reply]
If you want to run base models in AMLB, my recommendation is running them through AutoGluon, since it does the data preprocessing out of the box to vend to the models. This is the approach I do in TabRepo and the upcoming TabRepo 2.0.

Of course this is opinionated data preprocessing that might differ from other's implementations (and I'm biased ofc), but that will always be the case and you'd need to do similar logic yourself if you wanted it to be stand-alone in AMLB (aka, you'd need to build and maintain your own AutoML system to make the base models work as robustly as AutoML systems do). Not to mention ensuring they respect time limit and other constraints (num_cpus etc.)

**Comment by mfeurer** [comment]
+1

**Comment by fmohr** [comment]
Sorry for the super late reply. For what it's worth: I also consider this very reasonable. Personally, I think that there should actually be a difference between nominal and ordinal attributes even; not sure whether this has been discussed at some earlier point. With a centralized encoder, one takes an important decision away from the AutoML tool. Having seen the approach for these Riemann cells in PFNs, I am not sure whether the distinction between nominal and ordinal attributes actually matters in practice, but this is just a different type of attribute.

**Reply by PGijsbers** [reply]
Thanks for sharing! OpenML does not make a distinction between nominal and ordinal, so for the time being I will not work on support for it in AMLB either.

**Comment by Innixma** [comment]
+1. This always previously led to some oddities, such as when measuring inference speed, since the encoded input was able to get faster inference time because the actual time it takes for encoding wasn't counted for the model, whereas the non-encoded input had to be preprocessed by the AutoML frameworks. That is why TPOT's inference time looked crazy fast, since it didn't need to preprocess the data.

If you have to support the old frameworks, you can make a generic preprocessor that does the equivalent logic, which wraps the given AutoML frameworks so they continue to work.
